src/nnet3/nnet-simple-component.cc

```diff
@@ -1291,6 +1292,293 @@ void RepeatedAffineComponent::UnVectorize(const VectorBase<BaseFloat> &params) {
+                              linear_params_batch, kTrans, 1.0);
+
+  DeletePointers(&in_batch);
+  DeletePointers(&out_batch);
+  DeletePointers(&linear_params_batch);
+}
+
+void BlockAffineComponent::Backprop(const std::string &debug_info,
+                                    const ComponentPrecomputedIndexes *indexes,
+                                    const CuMatrixBase<BaseFloat> &in_value,
+                                    const CuMatrixBase<BaseFloat> &, // out_value
+                                    const CuMatrixBase<BaseFloat> &out_deriv,
+                                    Component *to_update_in,
+                                    CuMatrixBase<BaseFloat> *in_deriv) const {
+  BlockAffineComponent *to_update = dynamic_cast<BlockAffineComponent*>(to_update_in);
+
+  const int32 num_rows_in_block = linear_params_.NumRows() / num_blocks_;
+  const int32 num_cols_in_block = linear_params_.NumCols();
+
+  // Propagate the derivative back to the input.
+  // add with coefficient 1.0 since property kBackpropAdds is true.
+  // If we wanted to add with coefficient 0.0 we'd need to zero the
+  // in_deriv, in case of infinities.
+  if (in_deriv) {
+    std::vector<CuSubMatrix<BaseFloat> *> in_deriv_batch, out_deriv_batch, linear_params_batch;
+
+    for(int block_counter = 0; block_counter < num_blocks_; block_counter++) {
+      CuSubMatrix<BaseFloat> *in_deriv_block =
+        new CuSubMatrix<BaseFloat>(in_deriv->ColRange(block_counter * num_cols_in_block, 
+                                                      num_cols_in_block));
+      in_deriv_batch.push_back(in_deriv_block);
+
+      CuSubMatrix<BaseFloat> *out_deriv_block =
+        new CuSubMatrix<BaseFloat>(out_deriv.ColRange(block_counter * num_rows_in_block, 
+                                                       num_rows_in_block));
+      out_deriv_batch.push_back(out_deriv_block);
+
+      CuSubMatrix<BaseFloat> *linear_params_block =
+        new CuSubMatrix<BaseFloat>(linear_params_.RowRange(block_counter * num_rows_in_block,
+                                                          num_rows_in_block));
+      linear_params_batch.push_back(linear_params_block);
+    }
+
+    AddMatMatBatched<BaseFloat>(1.0, in_deriv_batch, out_deriv_batch, kNoTrans,
+                                linear_params_batch, kNoTrans, 1.0);
+
+    DeletePointers(&in_deriv_batch);
+    DeletePointers(&out_deriv_batch);
+    DeletePointers(&linear_params_batch);
+  }
+
+  if (to_update != NULL) {
+
+    { // linear params update
+
+      std::vector<CuSubMatrix<BaseFloat> *> in_value_batch,
+        out_deriv_batch, linear_params_deriv_batch;
+
+      // this constructor zeros out the entire matrix. Unfortunately, we need
+      // this for AddMatMatBatched even though beta is 0 because it is possible
+      // there may be NaN in the matrix's space, and 0 * NaN is NaN.
+      CuMatrix<BaseFloat> linear_params_deriv(linear_params_.NumRows(),
```

Comment:
I don't think the temporary is necessary here.  It was needed in the RepeatedAffineComponent to avoid a race condition, but here you can directly add to linear_params_.  The learning rate is normally large enough that any roundoff error is negligible, and anyway we do it like this elsewhere.